exercises/exercise-15/database.ts

```diff
@@ -1,13 +1,221 @@
+                let fieldValue = obj[field] as unknown as string;
+                let fieldWords = fieldValue.toLowerCase().split(' ');
+                for (let i = 0; i < words.length; i++) {
+                    if (check[i])
+                        continue;
+                    for (let fieldWord of fieldWords) {
+                        if (fieldWord == words[i]) {
+                            check[i] = true;
+                            break;
+                        }
+                    }
+                }
+            }
+            return !check.includes(false);
+        }
+        else {
+            for (let [key, value] of Object.entries(query)) {
+                let operator = value as Operators<T[keyof T]>;
+                let obj_key = key as keyof T;
+                if (instanceOfOperatorEq(operator)) {
+                    if (!(obj[obj_key] === operator.$eq))
+                        return false;
+                }
+                else if (instanceOfOperatorGt(operator)) {
+                    if (!(obj[obj_key] > operator.$gt))
+                        return false;
+                }
+                else if (instanceOfOperatorLt(operator)) {
+                    if (!(obj[obj_key] < operator.$lt))
+                        return false;
+                }
+                else if (instanceOfOperatorIn(operator)) {
+                    if (!(operator.$in.includes(obj[obj_key])))
+                        return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    protected handleOption(input: T[], option?: Option<T>): (typeof option) extends ProjectionOption<T> ? Partial<T>[] : T[] {
+        let result = input;
+        if (option == undefined)
+            return result;
+        if (instanceOfSortOption(option)) {
+            let sortRule = Object.entries(option.sort) as [keyof T, 1 | -1][];
+            result.sort((_a, _b) => {
+                for (let [k, v] of sortRule) {
+                    let a = _a, b = _b;
+                    if (v == -1) {
+                        [a, b] = [b, a];
+                    }
+                    if (a[k] < b[k])
+                        return -1;
+                    else if (a[k] > b[k])
+                        return 1;
+                }
+                return 0;
+            });
+        }
+        if (instanceOfProjectionOption(option)) {
+            let a = result.map((obj) => {
+                let keys = Object.keys(option.projection) as (keyof T)[];
+                let newobj = {} as Partial<T>;
+                for (let key of keys) {
+                    newobj[key] = obj[key];
+                }
+                return newobj;
+            });
+            return a as any; //https://github.com/microsoft/TypeScript/issues/24929
+        }
+        return result;
+    }
+
+    async find(query: Expression<T>, option?: Option<T>): Promise<(typeof option) extends ProjectionOption<T> ? Partial<T>[] : T[]> {
```

Comment:
`Partial` means all fields are optional while `projection` specifies the exact list of fields that will be in the result.